kubernetes/resource_kubernetes_pod_test.go

```diff
@@ -97,6 +97,19 @@ func TestAccKubernetesPod_initContainer_updateForcesNew(t *testing.T) {
 					resource.TestCheckResourceAttr("kubernetes_pod.test", "spec.0.init_container.0.command.3", "http://kubernetes.io"),
 					resource.TestCheckResourceAttr("kubernetes_pod.test", "spec.0.init_container.0.volume_mount.0.name", "workdir"),
 					resource.TestCheckResourceAttr("kubernetes_pod.test", "spec.0.init_container.0.volume_mount.0.mount_path", "/work-dir"),
+					resource.TestCheckResourceAttr("kubernetes_pod.test", "spec.0.dns_config.#", "1"),
+					resource.TestCheckResourceAttr("kubernetes_pod.test", "spec.0.dns_config.0.nameservers.#", "3"),
+					resource.TestCheckResourceAttr("kubernetes_pod.test", "spec.0.dns_config.0.nameservers.0", "1.1.1.1"),
+					resource.TestCheckResourceAttr("kubernetes_pod.test", "spec.0.dns_config.0.nameservers.1", "8.8.8.8"),
+					resource.TestCheckResourceAttr("kubernetes_pod.test", "spec.0.dns_config.0.nameservers.2", "9.9.9.9"),
+					resource.TestCheckResourceAttr("kubernetes_pod.test", "spec.0.dns_config.0.searches.#", "1"),
+					resource.TestCheckResourceAttr("kubernetes_pod.test", "spec.0.dns_config.0.searches.0", "kubernetes.io"),
+					resource.TestCheckResourceAttr("kubernetes_pod.test", "spec.0.dns_config.0.option.#", "2"),
+					resource.TestCheckResourceAttr("kubernetes_pod.test", "spec.0.dns_config.0.option.0.name", "ndots"),
+					resource.TestCheckResourceAttr("kubernetes_pod.test", "spec.0.dns_config.0.option.0.value", "1"),
+					resource.TestCheckResourceAttr("kubernetes_pod.test", "spec.0.dns_config.0.option.1.name", "use-vc"),
+					resource.TestCheckResourceAttr("kubernetes_pod.test", "spec.0.dns_config.0.option.1.value", ""),
```

Comment:
I've looked into preserving the _nilness_ of the optional string field `value` by defining it as a list of string with `MaxItems` set to 1 but this results in an awkward configuration syntax:

```
		dns_config {
			nameservers = ["1.1.1.1", "8.8.8.8", "9.9.9.9"]
			searches    = ["kubernetes.io"]

			option {
				name  = "ndots"
				value = [ 1 ]
			}

			option {
				name  = "inexistent-option-with-not-nil-empty-value"
				value = [ "" ]
			}

			option {
				name = "use-vc"
			}
		}
```

I failed to locate proper guidelines on how to implement this use case.

Anyway, I did not find any example of DNS config options which accept a value that is the empty string:
http://man7.org/linux/man-pages/man5/resolv.conf.5.html

And this specific case is eliminated by the kubelet implementation: 
https://github.com/kubernetes/kubernetes/blob/v1.14.0/pkg/kubelet/network/dns/dns.go#L298

Reply:
The provider SDK isn't able to distinguish `nil` in configuration as you would expect so that's not even worth considering.